From 2d8dffbbfc0c431a37319d4d9a143836c9ec542e Mon Sep 17 00:00:00 2001
From: Anton Bachin <antonbachin@yahoo.com>
Date: Sun, 12 Oct 2025 18:12:24 +0300
Subject: [PATCH] Adapt to Cmdliner 2.0.0

Bisect_ppx now effectively requires OCaml 4.08, because Cmdliner 2.0.0
requires it. However, the code of Bisect_ppx proper can still easily be
made to compile against OCaml 4.03.

Resolves #444.
---
 .github/workflows/test.yml |  7 +++----
 Makefile                   |  2 +-
 bisect_ppx.opam            |  2 +-
 src/report/main.ml         | 35 ++++++++++++-----------------------
 test/report/send-to.t      |  8 ++++----
 5 files changed, 21 insertions(+), 33 deletions(-)

diff --git a/src/report/main.ml b/src/report/main.ml
index 9f4f7191..37a2013a 100644
--- a/src/report/main.ml
+++ b/src/report/main.ml
@@ -21,29 +21,15 @@ let esy_source_dir =
   | exception Not_found -> []
   | directory -> [Filename.concat directory "default"]
 
-(* Many of the values used from Cmdliner.Term are deprecated in favor of values
-   from Cmdliner.Cmd. However, Cmdliner.Cmd was introduced in Cmdliner 1.1.0,
-   which requires OCaml 4.08.0. Bisect_ppx still supports OCaml 4.04, so Bisect
-   cannot use this recent version of Cmdliner. The 4.04 constraint itself is
-   only due to ppxlib.
-
-   So, suppress the deprecation warnings. *)
-module Term =
-struct
-  include Cmdliner.Term
-
-  let eval_choice = Cmdliner.Term.eval_choice [@ocaml.warning "-3"]
-  let exit = Cmdliner.Term.exit [@ocaml.warning "-3"]
-  let info = Cmdliner.Term.info [@ocaml.warning "-3"]
-end
-
 module Arg = Cmdliner.Arg
+module Cmd = Cmdliner.Cmd
+module Term = Cmdliner.Term
 
 
 
 (* Common arguments. *)
 
-let term_info = Term.info ~sdocs:"COMMON OPTIONS"
+let term_info = Cmd.info ~sdocs:"COMMON OPTIONS"
 
 let coverage_files from_position =
   Arg.(value @@ pos_right (from_position - 1) string [] @@
@@ -303,9 +289,8 @@ let merge =
 (* Entry point. *)
 
 let () =
-  Term.(eval_choice
-    (ret (const (`Help (`Auto, None))),
-    term_info
+  Cmd.group
+    (term_info
       "bisect-ppx-report"
       ~doc:"Generate coverage reports for OCaml and Reason."
       ~man:[
@@ -316,6 +301,10 @@ let () =
         `P
           ("See bisect-ppx-report $(i,COMMAND) --help for further " ^
           "information on each command, including options.")
-      ]))
-    [html; send_to; text; cobertura; coveralls; merge]
-  |> Term.exit
+      ]
+      ~exits:((Cmd.Exit.info ~doc:"on error." 1)::Cmd.Exit.defaults)
+      )
+    ([html; send_to; text; cobertura; coveralls; merge]
+    |> List.map (fun (term, info) -> Cmd.v info term))
+  |> Cmd.eval
+  |> exit
